datadog/reporter.go

```diff
@@ -0,0 +1,296 @@
+package datadog
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	// DefaultBufferSize contains the number of metrics the datadog reporter
+	// will capture before forcing a flush
+	DefaultBufferSize = 512
+)
+
+const (
+	typeGauge   = "gauge"
+	typeRate    = "rate"
+	typeCounter = "counter"
+	typeTimer   = "timer"
+)
+
+var (
+	poolMetric = &sync.Pool{
+		New: func() interface{} {
+			return &metric{}
+		},
+	}
+)
+
+type metric struct {
+	Name   string      `json:"metric"`
+	Points [][]float64 `json:"points"`
+	Type   string      `json:"type"`
+	Host   string      `json:"host,omitempty"`
+	Tags   []string    `json:"tags,omitempty"`
+}
+
+func (m *metric) Reset() {
+	m.Name = ""
+	m.Points = nil
+	m.Type = ""
+	m.Host = ""
+	m.Tags = nil
+}
+
+type series struct {
+	Metrics []*metric `json:"series"`
+}
+
+type Reporter struct {
+	metrics     []*metric
+	bufSize     int
+	offset      int
+	handlerFunc func(*http.Request) (*http.Response, error)
+	output      io.Writer
+	mux         *sync.Mutex
+	endpoint    string
+}
+
+// Reporting returns whether the reporter has the ability to actively report.
+func (r *Reporter) Reporting() bool {
+	return false
+}
+
+// Tagging returns whether the reporter has the capability for tagged metrics.
+func (r *Reporter) Tagging() bool {
+	return true
+}
+
+// Capabilities returns the capabilities description of the reporter.
+func (r *Reporter) Capabilities() tally.Capabilities {
+	return r
+}
+
+// post encoded metrics to datadog
+func (r *Reporter) post(data []byte) error {
+	req, err := http.NewRequest("POST", r.endpoint, bytes.NewReader(data))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := r.handlerFunc(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	io.Copy(r.output, resp.Body)
+	io.WriteString(r.output, "\n")
+
+	return nil
+}
+
+// submit the set of metrics to datadog. submit will retry the operation if
+// unsuccessful
+func (r *Reporter) submit(metrics []*metric) {
+	s := series{Metrics: metrics}
+
+	data, err := json.Marshal(s)
+	if err != nil {
+		return
+	}
+
+	for attempts := 0; attempts < 3; attempts++ {
+		if err := r.post(data); err != nil {
+			fmt.Fprintln(os.Stderr, "failed to connect to host")
+			fmt.Fprintln(r.output, "failed to connect to host")
```

Comment:
Perhaps make where this notification goes configurable?  Users may have their own logging library they wish to use, etc.